Sources/NIOResumableUpload/HTTPResumableUploadHandler.swift

```diff
@@ -0,0 +1,141 @@
+    public typealias InboundIn = HTTPRequestPart
+    public typealias InboundOut = Never
+
+    public typealias OutboundIn = Never
+    public typealias OutboundOut = HTTPResponsePart
+
+    var upload: HTTPResumableUpload
+
+    private var context: ChannelHandlerContext!
+    private var eventLoop: EventLoop!
+
+    /// Create an `HTTPResumableUploadHandler` within a given `HTTPResumableUploadContext`.
+    /// - Parameters:
+    ///   - context: The context for this upload handler.
+    ///              Use the same context across upload handlers, as uploads can't resume across different contexts.
+    ///   - channelConfigurator: A closure for configuring the child HTTP server channel.
+    public init(
+        context: HTTPResumableUploadContext,
+        channelConfigurator: @escaping (Channel) -> Void
+    ) {
+        self.upload = HTTPResumableUpload(
+            context: context,
+            channelConfigurator: channelConfigurator
+        )
+    }
+
+    /// Create an `HTTPResumableUploadHandler` within a given `HTTPResumableUploadContext`.
+    /// - Parameters:
+    ///   - context: The context for this upload handler.
+    ///              Use the same context across upload handlers, as uploads can't resume across different contexts.
+    ///   - handlers: Handlers to add to the child HTTP server channel.
+    public init(
+        context: HTTPResumableUploadContext,
+        handlers: [ChannelHandler] = []
+    ) {
+        self.upload = HTTPResumableUpload(context: context) { channel in
+            if !handlers.isEmpty {
+                _ = channel.pipeline.addHandlers(handlers)
+            }
+        }
+    }
+
+    public func handlerAdded(context: ChannelHandlerContext) {
+        self.context = context
+        self.eventLoop = context.eventLoop
+
+        self.upload.scheduleOnEventLoop(context.eventLoop)
+        self.upload.attachUploadHandler(self, channel: context.channel)
+    }
+
+    public func channelActive(context: ChannelHandlerContext) {
+        context.read()
+    }
+
+    public func channelInactive(context: ChannelHandlerContext) {
+        self.upload.end(handler: self, error: nil)
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        self.upload.receive(handler: self, channel: self.context.channel, part: unwrapInboundIn(data))
+    }
+
+    public func channelReadComplete(context: ChannelHandlerContext) {
+        self.upload.receiveComplete(handler: self)
+    }
+
+    public func channelWritabilityChanged(context: ChannelHandlerContext) {
+        self.upload.writabilityChanged(handler: self)
+    }
+
+    public func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {}
+
+    public func errorCaught(context: ChannelHandlerContext, error: Error) {
+        self.upload.end(handler: self, error: error)
+    }
+
+    public func read(context: ChannelHandlerContext) {
+        // Do nothing.
```

Comment:
Overriding this method to do nothing prevents upload resumption in HTTP/1.1 (the bug Marius was seeing), but I'm not entirely sure why. In the failing case, we do

```
HTTPResumableUpload offsetRetrieving(otherHandler:)
HTTPResumableUploadHandler channelRead(context:data:)
HTTPResumableUploadHandler channelReadComplete(context:)
HTTPResumableUploadHandler read(context:)
```

and the server hangs here while the client tries to send the PATCH request. If we remove this, the upload appending happens as expected right after offset retrieving. I'm still learning the NIO architecture, is there a reason we want to override the default `read(context:)` handling?

Reply:
This seems unlikely to be what we want. `read` is used to ask the channel to read more from the network, so doing this will suppress all reads. Probably not what we want.

In general, empty overrides should be removed: their default behaviour is better.

Reply:
There was a specific reason that we intercepted `read`. I think we wanted `read` to be driven by the child channel managed by the client's handlers, so this was the way to disable auto read behavior of the parent channel. Wondering if I missed something.

Reply:
That makes sense. Intercepting read like this does allow the child channel to drive the read behavior. But since the offset retrieving is transparent to the client's handler, I think the `read` doesn't propagate to the client's handler. I'm looking into how we could solve this while still overriding the `read`, maybe we need to explicitly attach the handler during offset retrieving?

Reply:
@Lukasa Guoye and I were debugging this and found that multiple H1 requests were being received on the same handler, which caused the issue since we were setting `self.uploadHandler` to `nil` (e.g. the subsequent upload appending `PATCH` wasn't being handled). Is there a way to ensure that only one request is delivered per handler?